msal/src/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs

```diff
@@ -0,0 +1,125 @@
+        private DeviceCodeResult _deviceCodeResult;
+
+        public DeviceCodeRequest(AuthenticationRequestParameters authenticationRequestParameters, Action<DeviceCodeResult> deviceCodeResultCallback) 
+            : base(authenticationRequestParameters)
+        {
+            _deviceCodeResultCallback = deviceCodeResultCallback;
+            
+            LoadFromCache = false;  // no cache lookup for token
+            SupportADFS = false;
+        }
+
+        internal override async Task PreTokenRequestAsync()
+        {
+            await base.PreTokenRequestAsync().ConfigureAwait(false);
+
+            OAuth2Client client = new OAuth2Client();
+
+            var deviceCodeScopes = new SortedSet<string>();
+            deviceCodeScopes.UnionWith(AuthenticationRequestParameters.Scope);
+            deviceCodeScopes.Add("offline_access");
+            deviceCodeScopes.Add("profile");
+            deviceCodeScopes.Add("openid");
+
+            client.AddBodyParameter(OAuth2Parameter.ClientId, AuthenticationRequestParameters.ClientId);
+            client.AddBodyParameter(OAuth2Parameter.Scope, deviceCodeScopes.AsSingleString());
+
+            // todo: THIS IS A MAJOR HACK.  Work with Shiung/Henrik on proper way to determine the device code endpoint
+            string deviceCodeEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint
+                .Replace("token", "devicecode")
+                .Replace("common", "organizations");
+
+            AuthenticationRequestParameters.Authority.TokenEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint.Replace("common", "organizations");
+
+            DeviceCodeResponse response = await client.ExecuteRequestAsync<DeviceCodeResponse>(
+                client.CreateFullEndpointUri(new Uri(deviceCodeEndpoint)), 
+                HttpMethod.Post,
+                AuthenticationRequestParameters.RequestContext).ConfigureAwait(false);
+
+            // todo: fix resource/scopes here for how we want to invoke this...
+            _deviceCodeResult = response.GetResult(AuthenticationRequestParameters.ClientId, "todo resource");
+            _deviceCodeResultCallback(_deviceCodeResult);
+        }
+
+        protected override async Task SendTokenRequestAsync()
+        {
+            TimeSpan timeRemaining = _deviceCodeResult.ExpiresOn - DateTimeOffset.UtcNow;
+
+            while (timeRemaining.TotalSeconds > 0.0)
+            {
+                try
+                {
+                    await base.SendTokenRequestAsync().ConfigureAwait(false);
+                    return;
+                }
+                catch (MsalServiceException ex)
+                {
+                    if (ex.ErrorCode.Equals("authorization_pending", StringComparison.OrdinalIgnoreCase))
+                    {
+                        await Task.Delay(TimeSpan.FromSeconds(_deviceCodeResult.Interval)).ConfigureAwait(false);

```

Comment:
>_deviceCodeResult.Interval [](start = 62, length = 26)

should probably evaluate the value of this before using it ... and perhaps we should have a min, and some "exponential" function to it?

Reply:
Exponential doesn't make sense here.  The user has a fixed time to respond with their device code login and the expectation is that once they've entered their code that they'll be signed in.  Exponential fallback here would present a very strange user experience where the user may have to wait longer than expected to complete their login experience.

Reply:
I agree on exponential, however I put "exponential" in my question to indicate something different from linear. but it might be fine with something linear for the duration (expected about 15 mins of polling). We should though still ensure that we have to deal with 429's correctly throughout the stack (currently expecting that will be lower in the stack than here). 

---
In reply to: [219945965](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1263#discussion_r219945965) [](ancestors = 219945965)